File: test/password_strength_test.rb
# frozen_string_literal: true

require 'test_helper'

class TestPasswordStrength < Minitest::Test
  def setup
    @strength = PasswordStrength::Base.new('johndoe', 'mypass')
    @password_limit = PasswordStrength::Base.const_get(:PASSWORD_LIMIT)
    @username_limit = PasswordStrength::Base.const_get(:USERNAME_LIMIT)
  end

  def teardown
    set_const(:PASSWORD_LIMIT, @password_limit)
    set_const(:USERNAME_LIMIT, @username_limit)
  end

  def test_shortcut
    @strength = PasswordStrength.test('johndoe', 'mypass')

    assert_kind_of PasswordStrength::Base, @strength
    assert_equal 'johndoe', @strength.username
    assert_equal 'mypass', @strength.password
  end

  def test_deal_with_empty_password
    @strength = PasswordStrength.test('johndoe', '')

    assert_predicate @strength, :weak?
  end

  def test_short_password
    @strength.password = 'xyz'
    @strength.test

    assert_equal 0, @strength.score
    assert_equal :weak, @strength.status
  end

  def test_password_equals_to_username
    @strength.password = 'johndoe'
    @strength.test

    assert_equal 0, @strength.score
    assert_equal :weak, @strength.status
  end

  def test_strong_password
    @strength.password = '^P4ssw0rd$'
    @strength.test

    assert_equal 100, @strength.score
    assert_equal :strong, @strength.status
  end

  def test_weak_password
    @strength.password = 'ytrewq'
    @strength.test

    assert_equal :weak, @strength.status

    @strength.password = 'asdfghjklm'
    @strength.test

    assert_equal :weak, @strength.status
  end

  def test_good_password
    @strength.password = '12345asdfg'
    @strength.test

    assert_equal :good, @strength.status

    @strength.password = '12345ASDFG'
    @strength.test

    assert_equal :good, @strength.status

    @strength.password = '12345Aa'
    @strength.test

    assert_equal :good, @strength.status
  end

  def test_reject_long_passwords_using_same_character
    @strength = PasswordStrength.test('johndoe', 'a' * 50)

    assert_predicate @strength, :invalid?
    assert_equal :repeated_character, @strength.invalid_reason
  end

  def test_exclude_option_as_regular_expression
    @strength = PasswordStrength.test('johndoe', '^Str0ng P4ssw0rd$', exclude: /\s/)

    assert_predicate @strength, :invalid?
    assert_equal :excluded_characters, @strength.invalid_reason
  end

  def test_exclude_option_as_array
    @strength = PasswordStrength.test('johndoe', 'asdfasdfasdf', exclude: %w[asdf 123])

    assert_predicate @strength, :invalid?
    assert_equal :excluded_characters, @strength.invalid_reason
  end

  def test_long_passwords_same_as_truncated
    set_const(:PASSWORD_LIMIT, 20)

    assert_equal(*[('ab' * 10), ('ab' * 100)].map { |password| result_for('johndoe', password) })
  end

  def test_long_usernames_same_as_truncated
    set_const(:USERNAME_LIMIT, 20)

    assert_equal(*[('ab' * 10), ('ab' * 100)].map { |username| result_for(username, '^Str0ng P4ssw0rd$') })
  end

  def set_const(const, value)
    PasswordStrength::Base.send(:remove_const, const)
    PasswordStrength::Base.const_set(const, value)
  end

  private

  # Everything the caller can read back from a test, so two runs can be
  # compared in one assertion.
  def result_for(username, password)
    strength = PasswordStrength.test(username, password)

    [strength.score, strength.password, strength.username, strength.status]
  end
end

File: test/js/password_strength.test.mjs
import assert from "node:assert/strict";
import { createRequire } from "node:module";
import { describe, it } from "node:test";

const require = createRequire(import.meta.url);
const PasswordStrength = require("../../app/assets/javascripts/password_strength.js");

describe("minimum length", () => {
  it("applies no minimum by default", () => {
    const strength = PasswordStrength.test("johndoe", "^P4ss$");

    assert.notEqual(strength.status, "invalid");
    assert.equal(strength.invalidReason, null);
  });

  it("rejects a password under the minimum", () => {
    const strength = PasswordStrength.test("johndoe", "^P4ssw0rd$", { minLength: 12 });

    assert.equal(strength.status, "invalid");
    assert.equal(strength.invalidReason, "too_short");
    assert.equal(strength.isValid("weak"), false);
  });

  it("accepts a password on the minimum", () => {
    const strength = PasswordStrength.test("johndoe", "^P4ssw0rd12$", { minLength: 12 });

    assert.equal(strength.status, "strong");
    assert.equal(strength.invalidReason, null);
  });

  it("counts characters rather than UTF-16 units", () => {
    const strength = PasswordStrength.test("johndoe", "çãéíóúàèìòù^1$", { minLength: 12 });

    assert.equal(strength.isTooShort(), false);
  });
});

describe("blocklist", () => {
  it("rejects a bundled common password", () => {
    const strength = PasswordStrength.test("johndoe", "password");

    assert.equal(strength.invalidReason, "common_word");
  });

  it("rejects a common password written in leet", () => {
    const strength = PasswordStrength.test("johndoe", "P@ssw0rd");

    assert.equal(strength.invalidReason, "common_word");
  });

  // Digits fold as well as symbols do: a class built with a backslash in front
  // of a digit reads as an octal escape and silently stops matching it.
  it("rejects a common password written with digits for letters", () => {
    const strength = PasswordStrength.test("johndoe", "l3tm31n");

    assert.equal(strength.invalidReason, "common_word");
  });

  it("rejects a common password followed by digits", () => {
    const strength = PasswordStrength.test("johndoe", "monkey2024");

    assert.equal(strength.invalidReason, "common_word");
  });

  it("accepts a passphrase that merely contains a common word", () => {
    const strength = PasswordStrength.test("johndoe", "blue-River-42-lamp");

    assert.equal(strength.isValid("good"), true);
  });

  it("uses a supplied list in place of the bundled one", () => {
    PasswordStrength.blocklist = ["tetherx"];

    try {
      assert.equal(PasswordStrength.test("johndoe", "T3therX99").invalidReason, "common_word");
      assert.notEqual(PasswordStrength.test("johndoe", "password").status, "invalid");
    } finally {
      PasswordStrength.blocklist = null;
    }
  });
});

describe("rejection reasons", () => {
  it("reports a repeated character", () => {
    const strength = PasswordStrength.test("johndoe", "a".repeat(50));

    assert.equal(strength.invalidReason, "repeated_character");
  });

  it("reports excluded characters", () => {
    const strength = PasswordStrength.test("johndoe", "^Str0ng P4ssw0rd$", { exclude: /\s/ });

    assert.equal(strength.invalidReason, "excluded_characters");
  });
});

describe("sequences", () => {
  it("counts a run of three", () => {
    const strength = PasswordStrength.test("johndoe", "mypass");

    assert.equal(strength.sequences("abc"), 1);
    assert.equal(strength.sequences("aaa"), 1);
  });

  it("does not read a character outside the basic plane as a run", () => {
    const strength = PasswordStrength.test("johndoe", "mypass");

    assert.equal(strength.sequences("😀"), 0);
  });
});

describe("parity with the Ruby scoring", () => {
  // The same corpus test/parity_test.rb asserts against, so a rule that changes
  // on one side and not the other fails here or there.
  const fixture = require("../fixtures/parity.json");

  for (const { password, status, reason } of fixture.cases) {
    it(`judges ${password} as ${status}`, () => {
      const strength = PasswordStrength.test(fixture.username, password, { minLength: fixture.min_length });

      assert.equal(strength.status, status);
      assert.equal(strength.invalidReason, reason);
    });
  }
});
